pkg/cmd/pulumi/about.go

```diff
@@ -0,0 +1,546 @@
+				Columns: []string{r.Type, r.URN},
+			}
+		}
+		resources = cmdutil.Table{
+			Headers: []string{"TYPE", "URN"},
+			Rows:    rows,
+		}.String() + "\n"
+	}
+	var pending string
+	if len(current.PendingOps) == 0 {
+		pending = fmt.Sprintf("Found no pending operations associated with %s\n", current.Name)
+	} else {
+		rows := make([]cmdutil.TableRow, len(current.PendingOps))
+		for i, r := range current.PendingOps {
+			rows[i] = cmdutil.TableRow{
+				Columns: []string{r.Type, r.URN},
+			}
+		}
+		pending = cmdutil.Table{
+			Headers: []string{"OPP TYPE", "URN"},
+			Rows:    rows,
+		}.String() + "\n"
+	}
+	return fmt.Sprintf("Current Stack: %s\n\n%s\n%s", current.Name, resources, pending)
+}
+
+func simpleTableRows(arr [][]string) []cmdutil.TableRow {
+	rows := make([]cmdutil.TableRow, len(arr))
+	for i, e := range arr {
+		rows[i] = cmdutil.TableRow{
+			Columns: e,
+		}
+	}
+	return rows
+}
+
+// This does not have an associated struct. It also does not make sense to
+// serialize.
+func formatProgramDependenciesAbout(language, root string) (string, error) {
+	var depInfo string
+	switch language {
+	case langNodejs:
+		depInfo = "package.json"
+	case langPython:
+		depInfo = "requirements.txt"
+	case "go":
+		depInfo = "go.mod"
+	case langDotnet:
+		return fmt.Sprintf("Please include the result of \"dotnet list package\""), nil
+	default:
+		return "", errors.New(fmt.Sprintf("Unknown Language: %s", language))
+	}
+
+	path := filepath.Join(root, depInfo)
+
+	return fmt.Sprintf("Please include the contents of \"%s\" in your report.\n", path), nil
+}
+
+type cliAbout struct {
+	Version    *semver.Version `json:"version"`
+	GoVersion  string          `json:"goVersion"`
+	GoCompiler string          `json:"goCompiler"`
+}
+
+func getCLIAbout() cliAbout {
+	var ver semver.Version
+	// Version is not supplied in test builds.
+	if version.Version != "" {
+		ver = semver.MustParse(version.Version)
+	}
+	return cliAbout{
+		Version:    &ver,
+		GoVersion:  runtime.Version(),
+		GoCompiler: runtime.Compiler,
+	}
+}
+
+func (cli cliAbout) String() string {
+	return cmdutil.Table{
+		Headers: []string{"CLI", ""},
+		Rows: simpleTableRows([][]string{
+			{"Version", cli.Version.String()},
+			{"Go Version", cli.GoVersion},
+			{"Go Compiler", cli.GoCompiler},
+		}),
+	}.String()
+}
+
+func formatLogAbout() string {
+	logDir := flag.Lookup("log_dir")
+	if logDir != nil && logDir.Value.String() != "" {
+		return fmt.Sprintf("Pulumi locates its logs in %s", logDir)
+	}
+	return fmt.Sprintf("Pulumi locates its logs in %s by default", os.TempDir())
+}
+
+type projectRuntimeAbout struct {
+	Language   string `json:"language"`
+	Executable string `json:"executable"`
+	// We want Version to conform to the semvar format: v0.0.0
+	Version string `json:"version"`
+}
+
+func getProjectRuntimeAbout(proj *workspace.Project) (projectRuntimeAbout, error) {
+	var ex, version string
+	var err error
+	var out []byte
+	language := proj.Runtime.Name()
+	switch language {
```

Comment:
This is fine for now, but just pointing out that this should probably live in the language host, i.e. the CLI could ask the language host for runtime information (as well as dependency information).

We have a similar TODO for `pulumi new` where we're basically using a switch on the lang, but the functionality should really live in the language host: https://github.com/pulumi/pulumi/issues/1334

https://github.com/pulumi/pulumi/blob/784c661c0cfb96f8fa609a45fc66735678dfa613/pkg/cmd/pulumi/new.go#L577

Maybe we should add a similar TODO here (and in `formatProgramDependenciesAbout` above)?

Reply:
That makes sense. I'll do that.